cannot concretize a spec with flags defined in the package spec AND in the compiler specification

### Steps to reproduce

simple concretize the following spec:
```
spack:
  config:
    concretizer: clingo
  concretizer:
    reuse: false
    unify: false
  specs:
  - bzip2%cce@15.0.0 ldflags="-L/opt/amdgpu/lib64/"

  compilers:
  - compiler:
      spec: cce@15.0.0
      paths:
        cc: cc
        cxx: CC
        f77: ftn
        fc: ftn
      flags:
        ldlibs: -ldl
      operating_system: rhel8
      target: x86_64
      modules:
      - PrgEnv-cray
      - cce/15.0.0
      environment:
        append_path:
          LD_LIBRARY_PATH: /opt/cray/pe/cce/15.0.0/cce/x86_64/lib
      extra_rpaths: []
```



### Error message

```
~/testEnv> spack concretize -f
==> Starting concretization
==> Error: {'-L/opt/amdgpu/lib64/'} does not equal set()
```

### Information on your system

* **Spack:** 0.20.0.dev0 (ef06b9db5bb2099dfd6161d763953612eec1314a)
* **Python:** 3.9.7
* **Platform:** darwin-monterey-cannonlake
* **Concretizer:** clingo

spack -d concretize -f
```
==> [2022-11-29-23:03:50.998257] build(bzip2)
==> [2022-11-29-23:03:50.998371] build(diffutils)
==> [2022-11-29-23:03:50.998402] build(libiconv)
Traceback (most recent call last):
  File "/Users/lroskop/spack/bin/spack", line 54, in <module>
    sys.exit(main())
  File "/Users/lroskop/spack/lib/spack/spack_installable/main.py", line 36, in main
    sys.exit(spack.main.main(argv))
  File "/Users/lroskop/spack/lib/spack/spack/main.py", line 1000, in main
    return _main(argv)
  File "/Users/lroskop/spack/lib/spack/spack/main.py", line 955, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "/Users/lroskop/spack/lib/spack/spack/main.py", line 983, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "/Users/lroskop/spack/lib/spack/spack/main.py", line 632, in _invoke_command
    return_val = command(parser, args)
  File "/Users/lroskop/spack/lib/spack/spack/cmd/concretize.py", line 45, in concretize
    concretized_specs = env.concretize(force=args.force, tests=tests)
  File "/Users/lroskop/spack/lib/spack/spack/environment/environment.py", line 1276, in concretize
    return self._concretize_separately(tests=tests)
  File "/Users/lroskop/spack/lib/spack/spack/environment/environment.py", line 1400, in _concretize_separately
    concretized_root_specs = spack.util.parallel.parallel_map(
  File "/Users/lroskop/spack/lib/spack/spack/util/parallel.py", line 135, in parallel_map
    raise_if_errors(*results, debug=debug)
  File "/Users/lroskop/spack/lib/spack/spack/util/parallel.py", line 83, in raise_if_errors
    raise RuntimeError(error_fmt.format(msg))
RuntimeError: [PID=22437] Traceback (most recent call last):
  File "/Users/lroskop/spack/lib/spack/spack/util/parallel.py", line 55, in __call__
    value = self.func(*args, **kwargs)
  File "/Users/lroskop/spack/lib/spack/spack/environment/environment.py", line 2268, in _concretize_task
    return _concretize_from_constraints(spec_constraints, tests)
  File "/Users/lroskop/spack/lib/spack/spack/environment/environment.py", line 2244, in _concretize_from_constraints
    return s.concretized(tests=tests)
  File "/Users/lroskop/spack/lib/spack/spack/spec.py", line 3062, in concretized
    clone.concretize(tests=tests)
  File "/Users/lroskop/spack/lib/spack/spack/spec.py", line 2978, in concretize
    self._new_concretize(tests)
  File "/Users/lroskop/spack/lib/spack/spack/spec.py", line 2953, in _new_concretize
    result = solver.solve([self], tests=tests)
  File "/Users/lroskop/spack/lib/spack/spack/solver/asp.py", line 2435, in solve
    result, _, _ = self.driver.solve(setup, specs, reuse=reusable_specs, output=output)
  File "/Users/lroskop/spack/lib/spack/spack/solver/asp.py", line 705, in solve
    answers = builder.build_specs(tuples)
  File "/Users/lroskop/spack/lib/spack/spack/solver/asp.py", line 2293, in build_specs
    self.reorder_flags()
  File "/Users/lroskop/spack/lib/spack/spack/solver/asp.py", line 2188, in reorder_flags
    assert spec_compiler_flags_set == compiler_flags_set, "%s does not equal %s" % (
AssertionError: {'-L/opt/amdgpu/lib64/'} does not equal set()

```

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

--- a/lib/spack/spack/solver/asp.py
+++ b/lib/spack/spack/solver/asp.py
@@ -2176,48 +2176,41 @@ class SpecBuilder(object):
         The solver determines wihch flags are on nodes; this routine
         imposes order afterwards.
         """
-        # nodes with no flags get flag order from compiler
         compilers = dict((c.spec, c) for c in all_compilers_in_config())
-        for pkg in self._flag_compiler_defaults:
-            spec = self._specs[pkg]
-            compiler_flags = compilers[spec.compiler].flags
-            for key in spec.compiler_flags:
-                spec_compiler_flags_set = set(spec.compiler_flags.get(key, []))
-                compiler_flags_set = set(compiler_flags.get(key, []))
-
-                assert spec_compiler_flags_set == compiler_flags_set, "%s does not equal %s" % (
-                    spec_compiler_flags_set,
-                    compiler_flags_set,
-                )
-
-                spec.compiler_flags[key] = compiler_flags.get(key, [])
-        # index of all specs (and deps) from the command line by name
         cmd_specs = dict((s.name, s) for spec in self._command_line_specs for s in spec.traverse())
 
-        # iterate through specs with specified flags
-        for key, sources in self._flag_sources.items():
-            pkg, flag_type = key
-            spec = self._specs[pkg]
-            compiler_flags = spec.compiler_flags.get(flag_type, [])
+        for spec in self._specs.values():
+            # if bootstrapping, compiler is not in config and has no flags
+            flagmap_from_compiler = {}
+            if spec.compiler in compilers:
+                flagmap_from_compiler = compilers[spec.compiler].flags
 
-            # order is determined by the DAG.  A spec's flags come after
-            # any from its ancestors on the compile line.
-            order = [s.name for s in spec.traverse(order="post", direction="parents")]
+            for flag_type in spec.compiler_flags.valid_compiler_flags():
+                from_compiler = flagmap_from_compiler.get(flag_type, [])
+                from_sources = []
 
-            # sort the sources in our DAG order
-            sorted_sources = sorted(sources, key=lambda s: order.index(s))
+                # order is determined by the  DAG. A spec's flags come after any of its ancestors
+                # on the compile line
+                source_key = (spec.name, flag_type)
+                if source_key in self._flag_sources:
+                    order = [s.name for s in spec.traverse(order="post", direction="parents")]
+                    sorted_sources = sorted(
+                        self._flag_sources[source_key], key=lambda s: order.index(s)
+                    )
 
-            # add flags from each source, lowest to highest precedence
-            flags = []
-            for source_name in sorted_sources:
-                source = cmd_specs[source_name]
-                extend_flag_list(flags, source.compiler_flags.get(flag_type, []))
+                    # add flags from each source, lowest to highest precedence
+                    for source_name in sorted_sources:
+                        source = cmd_specs[source_name]
+                        extend_flag_list(from_sources, source.compiler_flags.get(flag_type, []))
 
-            assert set(compiler_flags) == set(flags), "%s does not equal %s" % (
-                set(compiler_flags),
-                set(flags),
-            )
-            spec.compiler_flags.update({flag_type: source.compiler_flags[flag_type]})
+                # compiler flags from compilers config are lowest precedence
+                ordered_compiler_flags = from_compiler + from_sources
+                compiler_flags = spec.compiler_flags.get(flag_type, [])
+
+                msg = "%s does not equal %s" % (set(compiler_flags), set(ordered_compiler_flags))
+                assert set(compiler_flags) == set(ordered_compiler_flags), msg
+
+                spec.compiler_flags.update({flag_type: ordered_compiler_flags})
 
     def deprecated(self, pkg, version):
         msg = 'using "{0}@{1}" which is a deprecated version'

--- a/lib/spack/spack/test/concretize.py
+++ b/lib/spack/spack/test/concretize.py
@@ -325,6 +325,13 @@ class TestConcretize(object):
         assert set(client.compiler_flags["fflags"]) == set(["-O0", "-g"])
         assert not set(cmake.compiler_flags["fflags"])
 
+    def test_compiler_flags_from_compiler_and_dependent(self):
+        client = Spec("cmake-client %clang@12.2.0 platform=test os=fe target=fe cflags==-g")
+        client.concretize()
+        cmake = client["cmake"]
+        for spec in [client, cmake]:
+            assert spec.compiler_flags["cflags"] == ["-O3", "-g"]
+
     def test_concretize_compiler_flag_propagate(self):
         spec = Spec("hypre cflags=='-g' ^openblas")
         spec.concretize()

